libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * AP_Beacon_Marvelmind.cpp
+ *
+ *  Created on: 21.03.2017
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <termios.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <ctype.h>
+
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_Beacon_Marvelmind.h"
+
+extern const AP_HAL::HAL& hal;
+
+// constructor
+AP_Beacon_Marvelmind::AP_Beacon_Marvelmind(AP_Beacon &frontend,
+        AP_SerialManager &serial_manager) :
+        AP_Beacon_Backend(frontend) {
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon,
+            0);
+    if (uart != nullptr) {
+        uart->begin(
+                serial_manager.find_baudrate(
+                        AP_SerialManager::SerialProtocol_Beacon, 0));
+
+    //variables for the receive thread
+        hedge = new MarvelmindHedge();
+        if(hedge)
+        {
+            createMarvelmindHedge();
+            Parse_State = RECV_HDR; // current state of receive data
+            nBytesInBlockReceived = 0; // bytes received
+            dataId = 0;
+            startMarvelmindHedge();
+        }else{
+            //Initialising beacon failed
+        }
+    }
+
+}
+
+void AP_Beacon_Marvelmind::getOrAllocBeacon(struct StationaryBeaconPosition &b,
+        uint8_t address) {
+    uint8_t i;
+    uint8_t n_used = hedge->positionsBeacons.numBeacons;
+
+    if (n_used != 0)
+        for (i = 0; i < n_used; i++) {
+            if (hedge->positionsBeacons.beacons[i].address == address) {
+                b = hedge->positionsBeacons.beacons[i];
+                return;
+            }
+        }
+
+    if (n_used >= (MAX_STATIONARY_BEACONS - 1))
+        return;
+
+    hedge->positionsBeacons.numBeacons = (n_used + 1);
+    b = hedge->positionsBeacons.beacons[n_used];
+}
+
+void AP_Beacon_Marvelmind::process_beacons_positions_datagram(
+        struct StationaryBeaconPosition &b) {
+    uint8_t n = input_buffer[5]; // number of beacons in packet
+    uint8_t i, ofs;
+    uint8_t address;
+    int16_t x, y, z;
+
+    if ((1 + n * 8) != input_buffer[4])
+        return; // incorrect size
+
+    for (i = 0; i < n; i++) {
+        ofs = 6 + i * 8;
+
+        address = input_buffer[ofs + 0];
+        x = input_buffer[ofs + 1] | (((uint16_t) input_buffer[ofs + 2]) << 8);
+        y = input_buffer[ofs + 3] | (((uint16_t) input_buffer[ofs + 4]) << 8);
+        z = input_buffer[ofs + 5] | (((uint16_t) input_buffer[ofs + 6]) << 8);
+
+        getOrAllocBeacon(b, address);
+        if (b.address != NULL) {
```

Comment:
That why NULL is tricky ... address is uint8_t, so it just need to be b.address != 0 ^^